src/download.c

```diff
@@ -58,9 +58,12 @@ static struct list *failed = NULL;
  * traversing a manifest, doing hash comparisons, and (re)staging any files
  * whose hash miscompares.
  *
- * file->hash[0] acts as index into the arrays
+ * file->hash[0:1] acts as index into the arrays
  */
 #define SWUPD_CURL_HASH_BUCKETS 256
+#define HASH_VALUE(h) (h >= 'a' ? h - 'a' + 10 : h - '0')
+#define HASH_TO_KEY(hash) (HASH_VALUE(hash[0]) << 4 | HASH_VALUE(hash[1]))
```

Comment:
Maybe I'm being dense, but can you explain some of the magic here? Why are you shifting the first hash value left 4? Why add 10 to h - 'a' ?

Reply:
HASH_VALUE converts a character to its hexadecimal value (0 to 15 assuming that the input was from the set 0123456789abcdef ).
HASH_TO_KEY converts two characters to a value in the range 0 to 255.

Reply:
Exactly. I'll comment this macro.